### PR TITLE
fix  Clashing endpoint-powered pages break route sorting(#4038)

### PR DESCRIPTION
**Pull request by aolose**
issue #4038

fix:
1. `shadow` of shadowed page not change  after merge matching page/endpoint pairs   
2. fallthrough response  didn't working at client-side  （client-side don't know if the returned content is a fallthrough）
3. request from fallthrough shadowed page not match the right endpoint

update: [`0e123e6`](https://github.com/sveltejs/kit/pull/4139/commits/0e123e63623f089c7813b45c6a5c03c6437a640c)

There are multiple  fallthrough shadow page , like this:
`[a]  -- fallthrough -> [b] --fallthrough --> [c] `
Before you need to request three times from client-side,
now you just need send one request.

update: [`7397a20`](https://github.com/sveltejs/kit/pull/4139/commits/7397a202c46566d1dfbfac84df0b0d437ab853fd)
add test :
1. client-side  fallthrough from shadow page to another shadow page
2. client-side fallthrough from shadow page  to another shadow page then fallthrough  to no-shadow page

```
cd packages\kit\test\apps\basics
npx  playwright test -g 'Shadow fallthrough shadow'
```

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpx changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

File: packages/kit/src/core/dev/plugin.js [hunk 1]
```diff
@@ -105,7 +105,9 @@ export async function create_plugin(config, cwd) {
 						}),
 						routes: manifest_data.routes.map((route) => {
 							if (route.type === 'page') {
+								/**@type import('types').SSRPage**/
```
Comment (Rich-Harris) [inline packages/kit/src/core/dev/plugin.js:108]: `@type` doesn't do anything with `return`. You would need to declare a new variable or make it inline (`return /** @type {...} */ (...)`) but I don't think there's much value in it

```suggestion
```

File: packages/kit/src/runtime/client/renderer.js [hunk 2]
```diff
@@ -718,7 +718,7 @@ export class Renderer {
 			if (cached) return cached;
 		}
 
-		const [pattern, a, b, get_params, has_shadow] = route;
+		const [pattern, a, b, get_params, has_shadow, routeKey] = route;
```
Comment (Rich-Harris) [inline packages/kit/src/runtime/client/renderer.js:721]: removes the `has_shadow` flag, per comment above — would need corresponding change on line 772. note that we use snake_case, not camelCase

```suggestion
		const [pattern, a, b, get_params, shadow_key] = route;
```

File: packages/kit/src/runtime/client/renderer.js [hunk 3]
```diff
@@ -775,12 +775,12 @@ export class Renderer {
 						const res = await fetch(
 							`${url.pathname}${url.pathname.endsWith('/') ? '' : '/'}__data.json${url.search}`,
 							{
+								// @ts-ignore
 								headers: {
-									'x-sveltekit-load': 'true'
+									'x-sveltekit-load': routeKey
```
Comment (Rich-Harris) [inline packages/kit/src/runtime/client/renderer.js:780]:
```suggestion
									'x-sveltekit-load': shadow_key
```

File: packages/kit/src/runtime/client/renderer.js [hunk 4]
```diff
@@ -791,7 +791,7 @@ export class Renderer {
 									state: this.current
 								};
 							}
-
+							if (res.status === 204) return;
```
Comment (Rich-Harris) [inline packages/kit/src/runtime/client/renderer.js:794]: what is this for?

File: packages/kit/src/runtime/server/index.js [hunk 5]
```diff
@@ -151,6 +151,38 @@ export async function respond(request, options, state = {}) {
...
+											});
+										}
+									}
+									return response;
+								}
+							}
+						}
+					}
+				}
+
```
Comment (Rich-Harris) [inline packages/kit/src/runtime/server/index.js:185]: Could this not be replaced with something along these lines inside the `for` loop below?

```js
if (key && route.key !== key) continue;
```

File: packages/kit/types/internal.d.ts [hunk 6]
```diff
@@ -71,7 +71,14 @@ export type CSRComponent = any; // TODO
 
 export type CSRComponentLoader = () => Promise<CSRComponent>;
 
-export type CSRRoute = [RegExp, CSRComponentLoader[], CSRComponentLoader[], GetParams?, HasShadow?];
+export type CSRRoute = [
+	RegExp,
+	CSRComponentLoader[],
+	CSRComponentLoader[],
+	GetParams?,
+	HasShadow?,
```
Comment (Rich-Harris) [inline packages/kit/types/internal.d.ts:79]:
```suggestion

```

File: packages/kit/types/internal.d.ts [hunk 7]
```diff
@@ -211,6 +218,7 @@ export interface SSREndpoint {
 	type: 'endpoint';
 	pattern: RegExp;
 	params: GetParams;
+
```
Comment (Rich-Harris) [inline packages/kit/types/internal.d.ts:221]: not sure what these extra newlines are for?
```suggestion
```

File: packages/kit/types/internal.d.ts [hunk 8]
```diff
@@ -236,7 +244,9 @@ export interface SSROptions {
 	dev: boolean;
 	floc: boolean;
 	get_stack: (error: Error) => string | undefined;
+
```
Comment (Rich-Harris) [inline packages/kit/types/internal.d.ts:247]:
```suggestion
```

File: packages/kit/types/internal.d.ts [hunk 9]
```diff
@@ -236,7 +244,9 @@ export interface SSROptions {
...
 	handle_error(error: Error & { frame?: string }, event: RequestEvent): void;
+
```
Comment (Rich-Harris) [inline packages/kit/types/internal.d.ts:249]:
```suggestion
```

File: packages/kit/types/internal.d.ts [hunk 10]
```diff
@@ -247,10 +257,13 @@ export interface SSROptions {
 	};
 	prefix: string;
 	prerender: boolean;
+
```
Comment (Rich-Harris) [inline packages/kit/types/internal.d.ts:260]:
```suggestion
```

File: packages/kit/types/internal.d.ts [hunk 11]
```diff
@@ -247,10 +257,13 @@ export interface SSROptions {
...
 	read(file: string): Buffer;
+
```
Comment (Rich-Harris) [inline packages/kit/types/internal.d.ts:262]:
```suggestion
```

File: packages/kit/types/internal.d.ts [hunk 12]
```diff
@@ -247,10 +257,13 @@ export interface SSROptions {
...
 	root: SSRComponent['default'];
 	router: boolean;
 	service_worker?: string;
+
```
Comment (Rich-Harris) [inline packages/kit/types/internal.d.ts:266]:
```suggestion
```

File: packages/kit/types/internal.d.ts [hunk 13]
```diff
@@ -262,11 +275,13 @@ export interface SSROptions {
 		assets: string;
 		nonce: string;
 	}): string;
+
```
Comment (Rich-Harris) [inline packages/kit/types/internal.d.ts:278]:
```suggestion
```

**Review by Rich-Harris — CHANGES_REQUESTED** [review]
Thank you — I left comments inline. This PR will need some tests before we can merge it!

File: packages/kit/src/runtime/client/renderer.js (hunk 4 above)
Comment (aolose) [reply packages/kit/src/runtime/client/renderer.js:794]: When a request endpoint  from shadow page and endpoints return a fallthrough,
then client-side need to  know  the response is a fallthrough. 
We need a flag to tell the client-side that. so I return a 204 status to client-site.

File: packages/kit/src/runtime/server/index.js (hunk 5 above)
Comment (aolose) [reply packages/kit/src/runtime/server/index.js:185]: Yes, it's better.

**Review by Rich-Harris — CHANGES_REQUESTED** [review]
Thank you, but this will still need tests before it can be properly reviewed or merged

**Comment by Rich-Harris** [comment]
Can you undo the `pnpm-lock.yaml` changes please?

**Comment by Rich-Harris** [comment]
Thank you. I spent some time looking into this PR — unfortunately `master` has diverged quite a bit, so it isn't a straightforward merge. As explained in #4203 I think we need to use a simpler (if less sophisticated) implementation, since this is something of an edge case, so I'll close this in favour of that PR

**Comment by aolose** [comment]
> Thank you. I spent some time looking into this PR — unfortunately `master` has diverged quite a bit, so it isn't a straightforward merge. As explained in #4203 I think we need to use a simpler (if less sophisticated) implementation, since this is something of an edge case, so I'll close this in favour of that PR

I agree width you, less sophisticated  is better.

**Pull request closed without merging**